exercises/bracket-push/example.rs

```diff
@@ -0,0 +1,62 @@
+// The code below is a stub. Just enough to satisfy the compiler.
+// In order to pass the tests you can add-to or change any of this code.
+
+use std::collections::HashMap;
+pub struct Brackets {
+    all: Vec<char>,
+    pairs: HashMap<char, char>,
+}
+
+impl<'a> From<&'a str> for Brackets {
+    fn from(i: &str) -> Self {
+        Brackets::new(String::from(i), None)
+    }
+}
+
+impl Brackets {
+    pub fn new(s: String, pairs: Option<HashMap<char, char>>) -> Self {
+        let p = match pairs {
+            Some(x) => x,
+            None => Brackets::default(),
+        };
+
+        let valid = p.keys().chain(p.values()).collect::<Vec<_>>();
+        let all = s.chars().filter(|c| valid.contains(&c)).collect::<Vec<char>>();
+
+        Brackets {
+            all: all,
+            pairs: p.clone(),
+        }
+    }
+
+    fn default() -> HashMap<char, char> {
+        let mut d = HashMap::new();
+        d.insert('[', ']');
+        d.insert('{', '}');
+        d.insert('(', ')');
+        d
+    }
+
+    pub fn are_balanced(&self) -> bool {
+        let mut unmatched: Vec<char> = Vec::new();
+
+        for b in self.all.clone() {
```

Comment:
I needed to review https://doc.rust-lang.org/book/loops.html#for to answer why. because `for b in self.all` uses `IntoIterator` which consumes, whereas `self` is only borrowed. That implies that `for &b in self.all.iter()` will work, and I suggest it to not clone the vector.